<div class="clearfix" style="padding: 10px; padding-left: 0px">
<a href="http://bombora.com"><img src="https://app.box.com/shared/static/e0j9v1xjmubit0inthhgv3llwnoansjp.png" width="200px" class="pull-right" style="display: inline-block; margin: 5px; vertical-align: middle;"></a>
<h1> Bombora Data Science: <br> *Interview Exam* </h1>
</div>

<img width="200px" src="https://app.box.com/shared/static/15slg1mvjd1zldbg3xkj9picjkmhzpa5.png">

---
# Welcome

Welcome! This notebook contains interview exam questions referenced in the *Instructions* section in the `README.md`—please read that first, *before* attempting to answer questions here.

<div class="alert alert-info" role="alert" style="margin: 10px">
<p style="font-weight:bold">ADVICE</p>
<p>*Do not* read these questions, and panic, *before* reading the instructions in `README.md`.</p>
</div>

<div class="alert alert-warning" role="alert" style="margin: 10px">
<p style="font-weight:bold">WARNING</p>

<p>If using <a href="https://try.jupyter.org">try.jupyter.org</a> do not rely on the server for anything you want to last - your server will be <span style="font-weight:bold">deleted after 10 minutes of inactivity</span>. Save often and rember download notebook when you step away (you can always re-upload and start again)!</p>
</div>


## Have fun!

Regardless of outcome, getting to know you is important. Give it your best shot and we'll look forward to following up!

# Exam Questions

## 1. Algo + Data Structures

### Q 1.1: Fibionacci
![fib image](https://upload.wikimedia.org/wikipedia/commons/thumb/9/93/Fibonacci_spiral_34.svg/200px-Fibonacci_spiral_34.svg.png)

#### Q 1.1.1
Given $n$ where $n \in \mathbb{N}$ (i.e., $n$ is an integer and $n > 0$), write a function `fibonacci(n)` that computes the Fibonacci number $F_n$, where $F_n$ is defined by the recurrence relation:

$$ F_n = F_{n-1} + F_{n-2}$$

with initial conditions of:

$$ F_1 = 1,  F_2 = 1$$

#### Q 1.1.2
What's the complexity of your implementation?

#### Q 1.1.3
Consider an alternative implementation to compute Fibonacci number $F_n$ and write a new function, `fibonacci2(n)`.

#### Q 1.1.4
What's the complexity of your implementation?

#### Q 1.1.5
What are some examples of optimizations that could improve computational performance?


### Q 1.2: Linked List
![ll img](https://upload.wikimedia.org/wikipedia/commons/thumb/6/6d/Singly-linked-list.svg/500px-Singly-linked-list.svg.png)

#### Q 1.2.1
Consider a [singly linked list](https://en.wikipedia.org/wiki/Linked_list), $L$. Write a function `is_palindrome(L)` that detects if $L$ is a [palindrome](https://en.wikipedia.org/wiki/Palindrome), by returning a bool, `True` or `False`.


In [2]:
# class Node:
#   def __init__(self, val, next):
#     self.val = val
#     self.next = next

def is_palindrome(L) -> bool:
  # List to store Linked List elements
  L_arr = []

  # Traverse Linked List
  while L is not None:
    L_arr.append(L.val)
    L = L.next
  
  # Single line operation to compare list with its reverse (O(n))
  return L_arr == L_arr[::-1]



#### Q 1.2.2
What is the complexity of your implementation?

#### Time Complexity: $\mathcal{O}(n)$
 * Converting the linked list to a regular list needs traversing through all the elements - $\mathcal{O}(n)$.
 * Comparing the new list with its reversed list where reversing the list takes $\mathcal{O}(n)$ operations again.
 * Total  - $\mathcal{O}(2n) \sim \mathcal{O}(n)$.

#### Space Complexity: $\mathcal{O}(n)$
  * We create a new python list with $n$ elements in it, therefore $\mathcal{O}(n)$ .


#### Q 1.2.3
Consider an alternative implementation to detect if L is a palindrome and write a new function, `is_palindrome2(L)`.

In [3]:
def rec_fn(L, head):
  # Base case to stop at the last element of linked list using tail pointer L
  if L.next is None:
    # Compare tail pointer against the head and increment head. Return 1/0 if true/false
    if L.val == head.val:
      return head.next, 1
    else:
      return head.next, 0

  if L is not None:
    head, ans = rec_fn(L.next, head)
    # Compare tail pointer against head and increment head. Return 1/0 if true/false
    # Tail pointer is decremented from recursion stack
    # ans == 1 ensures True is returned only when all the elements are matched.
    if L.val == head.val:
      ans = ans == 1
      return head.next, ans
    else:
      return head.next, 0

# Flow starts here (Recurrence solution)
def is_palindrome2(L):
  # Head pointer to compare against tail pointer elements
  head = L
  # Go to the recursion function
  head, ans = rec_fn(L, head)
  return ans

#### Q 1.2.4
What's the complexity of this implementation?


#### Time Complexity: $\mathcal{O}(n)$
 * Recursion through the n elements of the linked list - $\mathcal{O}(n)$.
 * Total - $\mathcal{O}(n)$.

#### Space Complexity: $\mathcal{O}(n)$
  * The space complexity from the recursion stack is $\mathcal{O}(n)$ .


#### Q 1.2.5 
What are some examples of optimizations that could improve computational performance?


##### 1. The space complexity can be further improved to $\mathcal{O}(1)$ by the following steps: #####
  * Split the linked list into first half and second half.
  * Reverse the second half of linked list in place.
  * Compare first split with second split.


  

In [ ]:
# Pseudocode for optimization
def is_palindrom(L):
  first_half_tail = findFirstHalf(L)
  second_half_head = first_half_tail.next
  reverse_second_half(second_half_head)
  
  while second_half_head is not None:
    if L.val != second_half_head.val:
      return False
    L = L.next
    second_half_head = second_half_head.next

  return True


  

## 2. Prob + Stats

### Q 2.1: Finding $\pi$ in a random uniform?
<img src=https://www.epicurus.com/food/recipes/wp-content/uploads/2015/03/Pi-Day.jpg width="480">

Given a uniform random generator $[0,1)$ (e.g., use your language's standard libary to generate random value), write a a function `compute_pi` to compute [$\pi$](https://en.wikipedia.org/wiki/Pi).

### **Explanation** 

Assume the 2D points generated with uniform distribution lies within a square of side 1 unit. We inscribe a quarter circle (radius=1 unit) inside this square as shown below.

<img src="./uniform_dist.png" width="280" height="200">

To bring $\pi$ into the equation, we look into the ratio of area of the quarter-circle to the area of the square enclosing the quarter-circle.
$$
\begin{align}
\frac{Area\;of\;quarter\_circle}{Area\;of\;square} = \frac{\pi r^{2}/4}{r \times r} = \frac{\pi}{4}
\end{align}
$$

Here, $\pi$ is our unknown value. The monte-carlo methods tries to estimate the ratio of the two areas using the points from the random numbers generated.

Let, 
$$
pts\;inside\;quarter\;circle = k \\
total\;pts\;inside\;square = N \\
Area\;of\;quarter\_circle = A_c \\
Area\;of\;square = A_s \\
$$

To further identify if the generated point $i \in \mathcal{R}^{2}$ lies inside or outside the quarter-circle, we calculate the distance of the 2D point from the origin. This is used in calculation of $k$.

$$
k = \sum_{i=0}^N 1\;if\; x^2 + y^2 \leq 1\;else\;0

$$

Then, 
$$
\frac{A_c}{A_s} \approx \frac{k}{N} \\
$$

From equation (1)
$$
\begin{aligned}
\frac{\pi}{4} & \approx \frac{k}{N} \\
\pi & \approx 4 \times \frac{k}{N}
\end{aligned}
$$

The above equation approximates the value of $\pi$. As $N$ increases, the variance reduces and the equation is able to better approximate the value of $\pi$. In the program below, $N$ is chosen to be 1 million points.


In [11]:
import random

def compute_pi():
  # Generate 2D random number of the defined size
  N = 1000000
  points_2d = [[random.random() for i in range(2)] for j in range(N)]

  # For each point (x, y), calculate x^2 + y^2 and verify if it is inside the quarter circle
  # the calculated value should be less than or equal to radius of the circle i.e. 1
  # find total points inside quarter circle using this
  k = 0
  for i in range(len(points_2d)):
    dist = points_2d[i][0]**2 + points_2d[i][1]**2
    if dist <= 1:
      k += 1

  # area of quarter-circle/area of square ~ points inside quarter circle/total points
  pi = (4 * k)/N
  return pi

print(compute_pi())

3.141552


### Q 2.2: Making a 6-side die roll a 7?

Using a single 6-side die, how can you generate a random number between 1 - 7?

### Q 2.3: Is normality uniform?

<img src=https://rednaxela1618.files.wordpress.com/2014/06/uniformnormal.png width="480">


Given draws from a normal distribution with known parameters, how can you simulate draws from a uniform distribution?

### Q 2.4: Should you pay or should you go?

![coin flip](https://lh5.ggpht.com/iwD6MnHeHVAXNBgrO7r4N9MQxxYi6wT9vb0Mqu905zTnNlBciONAA98BqafyjzC06Q=w300)

Let’s say we play a game where I keep flipping a coin until I get heads. If the first time I get heads is on the nth coin, then I pay you $2^{(n-1)}$ US dollars. How much would you pay me to play this game? Explain.

### Q 2.5: Uber vs. Lyft

![uber vs lyft](http://usiaffinity.typepad.com/.a/6a01347fc1cb08970c01bb0876bcbe970d-pi)

You request 2 UberX’s and 3 Lyfts. If the time that each takes to reach you is IID, what is the probability that all the Lyfts arrive first? What is the probability that all the UberX’s arrive first?

### Q 2.6: Pick your prize
<img src=https://miro.medium.com/max/1100/1*m5b3O9sE68UCXjLw5oxy2g.png width="480">

A prize is placed at random behind one of three doors and you are asked to pick a door. To be concrete, say you always pick door 1. Now the game host chooses one of door 2 or 3, opens it and shows you that it is empty. They then give you the option to keep your picked door or switch to the unopened door. Should you stay or switch if you want to maximize your probability of winning the prize?

## 3 Conceptual ML

### Q 3.1 Why study gradient boosting or neural networks?

Consider a regression setting where $X \in \mathbb{R}^p$ and $Y \in \mathbb{R}$. The goal is to come up with a function $f(X): \mathbb{R}^p \rightarrow \mathbb{R}$ that minimizes the squared-error loss $(Y - f(X))^2$. Since X, Y are random variables, we seek to minimize the expectation of the squared error loss as follows
\begin{equation}
EPE(f) = \mathbb{E}\left[(Y-f(X)^2\right]
\end{equation}
where EPE stands for expected prediction error. One can show that minimizing the expected prediction error leads to the following _regression function_
\begin{equation}
f(x) = \mathbb{E}\left[Y|X=x\right]
\end{equation}

The goal of any method is to approximate the regression function above, which we denote as $\hat{f}(x)$. For example, linear regression explicitly assumes that the regression function is approximately linear in its arguments, i.e. $\hat{f}(x) = x^T\beta$ while a neural network provides a nonlinear approximation of the regression function. 

The simplest of all these methods is [k-nearest neighbors](https://en.wikipedia.org/wiki/K-nearest_neighbors_algorithm). Given $x$ and some neighbourhood of $k$ points $N_k(x)$, $\hat{f}(x)$ is simply the average of all $y_i|x_i \in N_k(x)$.  Let $N$ denote the training sample size. Under mild regularity conditions on the joint probability distribution $Pr(X, Y)$, one can show that as $N \rightarrow \infty$, $k \rightarrow \infty$ such that $k/N \rightarrow 0$, then $\hat{f}(x) \rightarrow f(x)$ where $\rightarrow$ means approaches or goes to. In other words, the k-nearest neighbors algorithm converges to the ideal solution as both the training sample size and number of neighbors increase to infinity.

Now given this _universal approximator_, why look any further and research other methods? Please share your thoughts.


#### Solution
The above mentioned universal approximator has a strong theoretical foundation which makes implicit assumptions regarding training size $N$ and the dimensionality $p$. Practical implementation of these assumptions are not feasible as the problem being solved gets more complicated. 

For example, the k-nearest neighbors algorithm is conditioned on a region and therefore, the output is always constant over that region. This leads to the fact that as $k$ grows in size, the algorithm becomes more stables. More specifically, as $N \rightarrow \infty$ and $k \rightarrow \infty, \hat{f}(x) \rightarrow \mathbb{E}[Y|X=x]$. In the real world, $N$ never reaches infinity and subsequently, $k$ cannot reach infinity. Coming up with large dataset for a complex problem (e.g. Video segmentation) is a infeasible. 

Additionally, as the dimension $p$ of the input grows in size, the data tends to be highly sparse. The notion that the points are geometrically close to each other belongs to the same probability distribution becomes invalid. Accordingly, the neighbors of a particular data point will be farther as compared to the lower dimensional counterpart. In this sense, the volume covered by the $k$ neighbors becomes very large to the extent that a small $k$ tend to cover the entire data space. 

The globally linear nature of linear regression functions and the additive modelling of these functions are another example of unrealistic assumptions that are used to tackle the dimensionality curse. Neural networks in general are shown to be universal approximator of any functions. It resolves the linearity assumption by including non-linear functionalities in its modelling such as sigmoid. Moreover, NNs have been shown to be successful in identifying low dimensional features that are generally not apparent in high dimensional space. 


### Q 3.2 Model Selection and Assesment

Consider a multiclass classification problem with a large number of features $p >> N$, for e.g $p=10000, N=100$ The task is threefold
1. Find a "good" subset of features that show strong _univariate_ correlation with class labels
2. Using the "good" subset, build a multi class classifier
3. Estimate the generalization error of the final model

Given this dataset, outline your approach and please be sure to cover the following
- Data splitting
- Model Selection: either estimating the performance of different classifiers or the same classifier with different hyperparameters
- Model Assessment: having chosen a classifier, estimating the generalization error

Assume all features are numerical, the dataset contains no NULLS, outliers, etc. and doesn't require any preprocessing.

